### PR TITLE
[Messenger] Don't prevent dispatch out of another bus

### PR DESCRIPTION
**Pull request by ogizanagi**
| Q             | A
| ------------- | ---
| Branch?       | master 
| Bug fix?      | no
| New feature?  | yes 
| Deprecations? | no 
| Tickets       | https://github.com/symfony/symfony/issues/35814#issuecomment-682433541 
| License       | MIT
| Doc PR        | N/A

### CODE REVIEW

File: src/Symfony/Component/Messenger/Middleware/DispatchAfterCurrentBusMiddleware.php [hunk 1]
```diff
@@ -44,12 +44,13 @@ class DispatchAfterCurrentBusMiddleware implements MiddlewareInterface
...
-                throw new \LogicException(sprintf('You can only use a "%s" stamp in the context of a message handler.', DispatchAfterCurrentBusStamp::class));
+            if ($this->isRootDispatchCallRunning) {
+                $this->queue[] = new QueuedEnvelope($envelope, $stack);
+
+                return $envelope;
             }
-            $this->queue[] = new QueuedEnvelope($envelope, $stack);
 
-            return $envelope;
+            $envelope = $envelope->withoutAll(DispatchAfterCurrentBusStamp::class);
```
Comment (ogizanagi) [inline src/Symfony/Component/Messenger/Middleware/DispatchAfterCurrentBusMiddleware.php:53]: Removing it from the envelope makes sense to me here for two reasons:
- even if it was explicitly configured with this stamp, it was not handled "transactionally". The userland bus might add it automatically by design and won't care about the context. So the stamp being present after the dispatch in the profiler might be confusing for the user.
- some logic could exist for messages dispatched transactionally with this stamp. Removing it in the case it was handled directly will prevent triggering such logic unexpectedly.

**Review by fabpot — APPROVED** [review]
LGTM. I would love to get some input from @Nyholm and/or @weaverryan

**Comment by Nyholm** [comment]
Thank you for the PR. 

Im hesitant. I do like the fact that this will make testing of my command handlers simpler because the command handler dispatches events with the `DispatchAfterCurrentBusStamp`. Now I don't need to invoke a dummy command bus that invoke my handlers, I can invoke my command handlers directly. 

However, when running production code (non-test), I don't see any scenario where you would like to add a DispatchAfter**CurrentBus**Stamp when you are not on a "current bus". So I dont think this change make any sense. 

With that said tough... If the stamp was named something differently. Like (not a suggestion) `DispatchMessageInTheirOwnSeparateBusTransaction`, then yes. I guess this would make perfect sense. =)

It also seams like @ostrolucky and @jkobus are dispatching these events from non-message handlers. That is something I would never do =). But I guess they have their reasons why it makes sense for them. So Im not sure I would fight them on a principle... 

----------

This answer was very undeceive. Sorry about that..

- I do like the implementation. 
- I don't like the reason why we need this
- I do like that this makes things easier for developers (myself included)
- I maybe just feel that merging this will result to a weird name of the stamp, but I don't suggest changing it. 

I can really go either way with this PR. @weaverryan what are your thoughts?

**Comment by fabpot** [comment]
Based on @Nyholm's comment, I'm merging it. Maybe another PR could change the name.

**Comment by fabpot** [comment]
Thank you @ogizanagi.

**Pull request merged**